File: 253_meeting_rooms_II.py
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e
# Real world problem, if there is no room available, open a new room; if there is, occupy it.
class Solution(object):
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        start = []
        end = []
        for i in intervals:
            start.append(i.start)
            end.append(i.end)
        start.sort()
        end.sort()
        # start and end pointer
        s = e = 0
        # number of rooms and available rooms
        numberOfRooms = availRooms = 0
        while s < len(start):
            if start[s] < end[e]:
                if availRooms == 0:
                    numberOfRooms += 1
                else:
                    availRooms -= 1
                s += 1
            else:
                availRooms += 1
                e += 1
        return numberOfRooms

File: 253_meeting_rooms_II.cc
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        map <int,int> m;
        for (auto i:intervals){
            m[i.start]++;
            m[i.end]--;
        }
        int total, res;
        total = res = 0;
        
        for (auto j:m){
            total += j.second;
            res = max(total,res);
        }
        return res;
        }
};
